drivers/usb/udc/udc_bflb_bl61x.c

```diff
@@ -0,0 +1,1480 @@
+	UBFBL61X_EVT_CTRL_END,
+	/* packet dma complete for specific endpoint */
+	UBFBL61X_EVT_END,
+};
+
+struct udc_bflb_bl61x_ev {
+	void *reserved;
+	const struct device *dev;
+	uint8_t ep_addr;
+	enum udc_bflb_bl61x_ev_type event;
+};
+
+K_MEM_SLAB_DEFINE(udc_bflb_bl61x_ev_slab, sizeof(struct udc_bflb_bl61x_ev),
+		  CONFIG_UDC_BFLB_BL61X_EVENT_COUNT, sizeof(void *));
+
+static void udc_bflb_bl61x_ev_submit(const struct device *dev,
+				 const uint8_t ep_addr,
+				 const enum udc_bflb_bl61x_ev_type event)
+{
+	struct udc_bflb_bl61x_data *priv = udc_get_private(dev);
+	struct udc_bflb_bl61x_ev *ev;
+	int ret;
+
+	ret = k_mem_slab_alloc(&udc_bflb_bl61x_ev_slab, (void **)&ev, K_NO_WAIT);
+	if (ret) {
+		udc_submit_event(dev, UDC_EVT_ERROR, ret);
+		LOG_ERR("Failed to allocate slab");
+		return;
+	}
+
+	ev->dev = dev;
+	ev->ep_addr = ep_addr;
+	ev->event = event;
+	k_fifo_put(&priv->fifo, ev);
+	k_work_submit_to_queue(udc_get_work_q(), &priv->work);
+}
+
+static enum udc_bus_speed udc_bflb_bl61x_device_speed(const struct device *dev)
+{
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
+	uint32_t speed;
+
+	speed = sys_read32(cfg->base + USB_OTG_CSR_OFFSET);
+	speed &= USB_SPD_TYP_HOV_POV_MASK;
+	speed = speed >> USB_SPD_TYP_HOV_POV_SHIFT;
+
+	if (speed == USB_BL61X_SPEED_FULL) {
+		return UDC_BUS_SPEED_FS;
+	} else if (speed == USB_BL61X_SPEED_HIGH) {
+		return UDC_BUS_SPEED_HS;
+	} else {
+		return UDC_BUS_UNKNOWN;
+	}
+
+	return UDC_BUS_UNKNOWN;
+}
+
+static uint8_t udc_bflb_bl61x_fifo_get_ep(const struct device *dev, uint8_t fifo)
+{
+	if (udc_bflb_bl61x_device_speed(dev) == UDC_BUS_SPEED_FS) {
+		return fifo;
+	}
+
+	if (fifo < 2) {
+		return 1;
+	}
+	return 2;
+}
+
+static void udc_bflb_bl61x_ctrl_ack(const struct device *dev)
+{
+	uint32_t tmp;
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
+
+	tmp = sys_read32(cfg->base + USB_DEV_CXCFE_OFFSET);
+	tmp |= USB_CX_DONE;
+	sys_write32(tmp, cfg->base + USB_DEV_CXCFE_OFFSET);
+}
+
+static void udc_bflb_bl61x_ep_ack(const struct device *dev, const uint8_t ep_idx)
+{
+	uint32_t tmp;
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
+
+	tmp = sys_read32(cfg->base + USB_DEV_INMPS1_OFFSET + (ep_idx - 1) * USB_BL61X_XPS_X_OFFSET);
+	tmp |= USB_TX0BYTE_IEP1;
+	sys_write32(tmp, cfg->base + USB_DEV_INMPS1_OFFSET + (ep_idx - 1) * USB_BL61X_XPS_X_OFFSET);
+}
+
+static uint8_t udc_bflb_bl61x_epcaps_to_eptype(const struct udc_ep_caps * const caps)
+{
+	uint8_t ret = 0;
+
+	if (caps->interrupt) {
+		ret |= USB_BL61X_CAP_INT;
+	}
+	if (caps->bulk) {
+		ret |= USB_BL61X_CAP_BULK;
+	}
+	if (caps->iso) {
+		ret |= USB_BL61X_CAP_ISO;
+	}
+	return ret;
+}
+
+static void udc_bflb_bl61x_fifo_configure(const struct device *dev,
+					uint8_t fifo_idx,
+					uint8_t ep_type,
+					uint16_t block_size,
+					uint8_t block_num,
+					bool enabled)
+{
+	const struct udc_bflb_bl61x_config *cfg = dev->config;
+	uint32_t tmp;
+
+	if (fifo_idx > 4) {
+		return;
+	}
```

Comment:
If it is an invalid configuration, as an alternative:
```suggestion
	__ASSERT_NO_MSG(fifo_idx <= 4);
```

Reply:
done